src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,267 @@
+            INamedTypeSymbol cancellationTokenType,
+            [NotNullWhen(returnValue: true)] out string? cancellationTokenParameterName)
+        {
+            cancellationTokenParameterName = null;
+
+            IMethodSymbol method = invocation.TargetMethod;
+
+            // Verify that the current invocation is not passing a token already
+            if (invocation.Arguments.Any(a => a.Parameter.Type.Equals(cancellationTokenType) && !a.IsImplicit))
+            {
+                return false;
+            }
+
+            // Check if the invocation's method has either an optional implicit ct not being used or a params ct parameter not being used or an overload that takes a ct
+            if (!InvocationMethodTakesAToken(method, invocation.Arguments, cancellationTokenType) &&
+                !MethodHasCancellationTokenOverload(method, cancellationTokenType))
+            {
+                return false;
+            }
+
+            // Check if the ancestor method has a ct that we can pass to the invocation
+            if (!VerifyAncestorOnlyHasOneCancellationTokenAsLastArgument(cancellationTokenType, GetContainingSymbol(invocation, containingSymbol), out cancellationTokenParameterName))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        // Try to find the most immediate containing symbol (anonymous or local function). If none is found, return the context containing symbol.
+        private static IMethodSymbol GetContainingSymbol(IInvocationOperation invocation, IMethodSymbol containingSymbol)
+        {
+            IOperation currentOperation = invocation.Parent;
+
+            while (currentOperation != null)
+            {
+                if (currentOperation.Kind == OperationKind.AnonymousFunction)
+                {
+                    return ((IAnonymousFunctionOperation)currentOperation).Symbol;
+                }
+                else if (currentOperation.Kind == OperationKind.LocalFunction)
+                {
+                    return ((ILocalFunctionOperation)currentOperation).Symbol;
+                }
+
+                currentOperation = currentOperation.Parent;
+            }
+
+            return containingSymbol;
+        }
+
+        // Check if the method only takes one ct and is the last parameter in the method signature.
+        // We want to compare the current method signature to any others with the exact same arguments in the exact same order.
+        private static bool VerifyAncestorOnlyHasOneCancellationTokenAsLastArgument(
+            INamedTypeSymbol cancellationTokenType,
+            IMethodSymbol methodDeclaration,
+            [NotNullWhen(returnValue: true)] out string? cancellationTokenParameterName)
+        {
+            if (methodDeclaration.Parameters.Count(x => x.Type.Equals(cancellationTokenType)) == 1 &&
+                methodDeclaration.Parameters.Last() is IParameterSymbol lastParameter &&
```

Comment:
Suggesting to use the indexer access as you are doing so on line 189.

```suggestion
                methodDeclaration.Parameters[methodDeclaration.Parameters.Length - 1] is IParameterSymbol lastParameter &&
```

Reply:
I updated all the `Last()` invocations to use the accessor. I made sure they all expect at least one item in the array.